### PR TITLE
Fix bad ipv6 comparison

### PR DESCRIPTION
**Pull request by lhh**
When performing DNS lookups, the source address from resolv.conf
may have stray zeroes in it, or the address string returned from
socket.recvfrom may. Purge them before comparing tuples.

Resolves: rhbz#1607967

Signed-off-by: Lon Hohberger <lon@metamorphism.com>

### CODE REVIEW

**Comment by lhh** [comment]
The source is this bug:

https://bugzilla.redhat.com/show_bug.cgi?id=1607967

**Comment by lhh** [comment]
I can - but I need to head out for the day - Monday?

**Comment by temoto** [comment]
@lhh sure, take your time.

**Comment by lhh** [comment]
I found an infinite loop as well (on source address mismatch). I'm not sure if this is how you'd like the tests to be written; I wanted to specifically test some of the logic flows in the udp() function.

**Comment by lhh** [comment]
That's an odd failure case which appears unrelated to these commits; @temoto - can you retrigger it?

The rest of the tests, particularly the new ones, passed:
test_udp_ipv6 (tests.greendns_test.TestUdp) ... ok
test_udp_ipv6_addr_zeroes (tests.greendns_test.TestUdp) ... ok
test_udp_ipv6_timeout (tests.greendns_test.TestUdp) ... ok
test_udp_ipv6_wrong_addr (tests.greendns_test.TestUdp) ... ok
test_udp_ipv6_wrong_addr_ignore (tests.greendns_test.TestUdp) ... ok

File: tests/greendns_test.py [hunk 1]
```diff
@@ -223,6 +223,104 @@ def getaliases(self, *args, **kwargs):
...
+    def test_udp_ipv6_wrong_addr(self):
+        _mock_socket.no_reply = False
+        # Should catch UnexpectedSource
+        _mock_socket.packet_source = ('ffff::1', 53, 0, 0)
+        result = False
+        try:
+            greendns.udp(self.query, '::1')
+        except dns.query.UnexpectedSource:
+            result = True
+        assert result
```
Comment (temoto) [inline tests/greendns_test.py]:
```
with tests.assert_raises(dns.query.UnexpectedSource):
  greendns.udp(self.query, '::1')
```
And similar in sibling tests for Timeout.

File: tests/greendns_test.py [hunk 2]
```diff
@@ -223,6 +223,104 @@ def getaliases(self, *args, **kwargs):
...
+        greendns.socket = self.orig_socket
+
+    def test_udp_ipv6(self):
+        _mock_socket.no_reply = False
+        greendns.udp(self.query, '::1')
+
+    def test_udp_ipv6_timeout(self):
+        _mock_socket.no_reply = True
+        try:
+            greendns.udp(self.query, 'a::1', timeout=0.1)
```
Comment (temoto) [inline tests/greendns_test.py]:
```
-class _mock_socket
-class _mock_socket_module

with mock.patch('greendns.socket.socket.recvfrom', side_effect=socket.timeout):
  greendns.udp(self.query, '::1')
```

**Comment by lhh** [comment]
Last commit had a sendto() that wasn't ipv6 loopback, so it failed.

**Comment by lhh** [comment]
Failure was in unrelated test:

https://travis-ci.org/eventlet/eventlet/jobs/410446056#L692

**Comment by temoto** [comment]
@lhh yes don't mind unrelated tests. Thank you very much for test.

**Pull request merged**